pkg/micro/client.go

```diff
@@ -183,6 +181,130 @@ func (c *Client) Request(
+	endpoint string,
+	payload proto.Message,
+	eventSubject string,
+	opts ...RequestAndSubscribeOption,
+) (*nats.Msg, error) {
+	// Apply default timeouts
+	cfg := &requestAndSubscribeConfig{
+		requestTimeout:  defaultRequestTimeout,
+		responseTimeout: defaultResponseTimeout,
+	}
+
+	// Apply custom options
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	return c.requestAndSubscribeWithTimeouts(
+		ctx, address, endpoint, payload, eventSubject,
+		cfg.requestTimeout, cfg.responseTimeout,
+	)
+}
+
+// requestAndSubscribeWithTimeouts is the internal implementation of RequestAndSubscribe.
+func (c *Client) requestAndSubscribeWithTimeouts(
+	ctx context.Context,
+	address *ServiceAddress,
+	endpoint string,
+	payload proto.Message,
+	eventSubject string,
+	requestTimeout time.Duration,
+	responseTimeout time.Duration,
+) (*nats.Msg, error) {
+	// Warn if parent context deadline is shorter than the combined internal timeouts.
+	// This helps developers identify potential timeout issues early.
+	totalInternalTimeout := requestTimeout + responseTimeout
+	if deadline, ok := ctx.Deadline(); ok {
+		remaining := time.Until(deadline)
+		if remaining < totalInternalTimeout {
+			c.log.Warn().
+				Str("endpoint", endpoint).
+				Str("event_subject", eventSubject).
+				Dur("context_remaining", remaining).
+				Dur("request_timeout", requestTimeout).
+				Dur("response_timeout", responseTimeout).
+				Dur("total_internal_timeout", totalInternalTimeout).
+				Msg("Parent context deadline is shorter than internal timeouts, operation may timeout early")
+		}
+	}
+
+	// Step 1: Prepare the request payload
+	anyPayload, err := anypb.New(payload)
+	if err != nil {
+		return nil, eris.Wrapf(ErrFailedToCreatePayload, "%v", err)
+	}
+
+	req := &microv1.Request{
+		ServiceAddress: address,
+		Payload:        anyPayload,
+	}
+
+	reqBytes, err := proto.Marshal(req)
+	if err != nil {
+		return nil, eris.Wrapf(ErrFailedToMarshal, "%v", err)
+	}
+
+	// Subscribe to event subject BEFORE sending request to prevent race condition.
+	// Using SubscribeSync (not Subscribe) to allow NextMsgWithContext for timeout control.
+	sub, err := c.SubscribeSync(eventSubject)
+	if err != nil {
+		return nil, eris.Wrapf(ErrFailedToSubscribe, "%v", err)
+	}
+
+	// Ensure cleanup in all paths (errors, timeouts, and success)
+	defer func() {
+		if err := sub.Unsubscribe(); err != nil {
+			c.log.Warn().Err(err).Str("subject", eventSubject).Msg("Failed to unsubscribe")
+		}
+	}()
+
+	// Auto-cleanup after one message for the success path.
+	if err := sub.AutoUnsubscribe(1); err != nil {
+		return nil, eris.Wrapf(ErrFailedToAutoUnsubscribe, "%v", err)
+	}
+
+	// Send command with immediate validation timeout.
+	// Using RequestWithContext (not Publish) because handler calls msg.Respond() for validation.
+	requestCtx, requestCancel := context.WithTimeout(ctx, requestTimeout)
+	defer requestCancel()
+
+	_, err = c.RequestWithContext(requestCtx, Endpoint(address, endpoint), reqBytes)
```

Comment:
The `micro.Response` here could contain errors (e.g. invalid command payload). We should just call `Request` here and only fetch the event from the subscription if the command is handled successfully.